### PR TITLE
PHP Stories API for Centralised Data Querying

### PR DESCRIPTION
**Pull request by ankitrox**
## Summary

- Adds a Stories class that can be used by users to list stories in different view formats based on the attributes provided.

## Relevant Technical Choices

- `Stories` class which uses the Renderer classes to display stories based on given attributes.
- `GenericRenderer` class which handles the circles, grid, list, and carousel story view types.
- `CarouselRenderer` class to render stories in carousel view.

## Testing Instructions

Displaying stories with default arguments.
```php
$stories = new Stories();
echo $stories->render();
```

Displaying Stories with user-provided arguments.
```php
$story_attributes = [
	'view_type'         => 'carousel',
	'show_story_poster' => false,
	'show_title'        => true,
];

$query_arguments = [
	'posts_per_page' => 10,
	'order'          => 'ASC',
	'orderby'        => 'title',
];
$stories = new Story_Query( $story_attributes, $query_arguments );
echo $stories->render( [ 'height' => 485, 'width' => 320 ] );
```

### CODE REVIEW

File: includes/Stories_Renderer/Fields/BaseField.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+	/**
+	 * BaseField constructor.
+	 *
+	 * @param array $args Arguments.
+	 */
+	public function __construct( array $args ) {
+		$this->label    = isset( $args['label'] ) ? $args['label'] : '';
+		$this->readonly = isset( $args['readonly'] ) ? (bool) $args['readonly'] : true;
+		$this->show     = true;
```
Comment (spacedmonkey) [inline includes/Stories_Renderer/Fields/BaseField.php:63]: What is the point of this property if it is always true?

Comment (spacedmonkey) [reply includes/Stories_Renderer/Fields/BaseField.php:63]: Wouldnt it be better to 
```
private $show = true;
```

File: tests/phpunit/tests/Stories_Renderer/Fields/BaseField.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	 */
+	public function test_label() {
+		$this->assertSame( 'Test Label', self::$testee->label() );
+	}
+
+	/**
+	 * @covers ::readonly
+	 */
+	public function test_readonly() {
+		$this->assertSame( true, self::$testee->readonly() );
```
Comment (spacedmonkey) [inline tests/phpunit/tests/Stories_Renderer/Fields/BaseField.php:59]:
```suggestion
		$this->assertTrue, self::$testee->readonly() );
```

File: tests/phpunit/tests/Stories_Renderer/Fields/BaseField.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	/**
+	 * @covers ::show
+	 */
+	public function test_show() {
+		$this->assertSame( true, self::$testee->show() );
```
Comment (spacedmonkey) [inline tests/phpunit/tests/Stories_Renderer/Fields/BaseField.php:66]:
```suggestion
		$this->assertTrue( self::$testee->show() );
```

File: tests/phpunit/tests/Story_Query.php [hunk 4]
```diff
@@ -1,4 +1,13 @@
 <?php
+/**
```
Comment (spacedmonkey) [inline tests/phpunit/tests/Story_Query.php:2]: What is happening in this file? This doesn't seem like a valid test anymore.

File: includes/Stories_Renderer/Renderer.php [hunk 5]
```diff
@@ -182,6 +187,34 @@ public function init() {
...
+				$field_state = new ListView();
+				break;
+			case 'circles':
+				$field_state = new CircleView();
+				break;
+			case 'carousel':
+				$field_state = new CarouselView();
+				break;
+			default:
+				$field_state = apply_filters( 'web_stories_default_fieldstate', new GridView() );
```
Comment (spacedmonkey) [inline includes/Stories_Renderer/Renderer.php:212]: We should have a filter comment, similar to 
https://github.com/google/web-stories-wp/blob/782c4029d28b938280c80203ca8d3b222a993c0c/includes/Traits/Publisher.php#L92-L103

**Comment by spacedmonkey** [comment]
Without tests for story_Query.php it is hard to know what is going on here.

**Comment by ankitrox** [comment]
> or story_Query.php it is hard to know what is going on here.

I will check the `Story_Query` tests and add it for re-review.

File: includes/Stories_Renderer/Carousel_Renderer.php [hunk 6]
```diff
@@ -96,7 +96,7 @@ class="<?php echo esc_attr( $container_classes ); ?>"
 					layout="intrinsic"
 					type="carousel"
 					role="region"
-					aria-label="<?php esc_attr_e( 'Web Stories', 'web-stories' ); ?>"
+					aria-label="<?php esc_attr_e( 'Basic carousel', 'web-stories' ); ?>"
```
Comment (swissspidy) [inline includes/Stories_Renderer/Carousel_Renderer.php:99]: Need to undo this change

File: tests/phpunit/tests/Stories_Renderer/Fields/BaseField.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+				'readonly' => true,
+			]
+		);
+	}
+
+	/**
+	 * @covers ::__construct
+	 */
+	public function test_instance() {
+		$this->assertInstanceOf( Testee::class, self::$testee );
```
Comment (swissspidy) [inline tests/phpunit/tests/Stories_Renderer/Fields/BaseField.php:45]: Pointless test?

File: includes/Stories_Renderer/Renderer.php [hunk 8]
```diff
@@ -182,6 +187,44 @@ public function init() {
...
+				$field_state = new CarouselView();
+				break;
+			default:
+				$default_field_state = new CircleView();
+				/**
+				 * Filters the fieldstate object.
+				 *
+				 * This depicts
+				 *
+				 * @since 2.0.0
```
Comment (spacedmonkey) [inline includes/Stories_Renderer/Renderer.php:218]: For now, let's go with 1.3.0

File: includes/functions.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+namespace Google\Web_Stories;
+
+/**
+ * Fetch stories based on customizer settings.
+ *
+ * @param array $args Arguments for fetching stories.
+ *
+ * @return string|void
+ */
+function stories( $args = [] ) {
```
Comment (spacedmonkey) [inline includes/functions.php:33]: I think we want to understand what we will make public api and what will not. Would love the thoughts on @swissspidy on this one.

**Review by spacedmonkey — APPROVED** [review]
Approved with one nit pic ( version number ) and one question for Pascal.

File: includes/functions.php (hunk 9 above)
Comment (ankitrox) [reply includes/functions.php:33]: @spacedmonkey  This is just a minimal version (as placeholder) for template tag. We need to discuss over its behaviour.

**Comment by ankitrox** [comment]
@spacedmonkey Nit pic addressed. Merging this branch.
